### PR TITLE
TypeID fix.

### PR DESCRIPTION
**Pull request by dkorolev**
```
[----------] Global test environment tear-down
[==========] 373 tests from 57 test cases ran. (8650 ms total)
[  PASSED  ] 373 tests.

ALL TESTS PASS.
```

### CODE REVIEW

File: TypeSystem/Reflection/reflection.h [hunk 1]
```diff
@@ -149,18 +143,24 @@ struct ReflectorImpl {
...
+        // Mark this structure as being visited, to make sure self-referring structs are correctly supported.
+        // The `incomplete_id` can really be anything as long as it's unique and different per type.
+        const TypeID incomplete_id =
+            static_cast<TypeID>(TYPEID_INCOMPLETE_STRUCT_TYPE +
+                                std::hash<std::type_index>()(std::type_index(typeid(T))) % TYPEID_TYPE_RANGE);
+        s.type_id = incomplete_id;
+
+        // As this structure has been marked used, visit all its fields recursively.
+        // StructFieldReflector<T> field_names_reflector(s.fields, true);
+        // VisitAllFields<T, FieldTypeAndNameAndIndex>::WithoutObject(field_names_reflector);
```
Comment (mzhurovich) [inline TypeSystem/Reflection/reflection.h]: Hmm. Is this some legacy from the refactoring process?

File: TypeSystem/Reflection/types.h [hunk 2]
```diff
@@ -186,27 +186,16 @@ inline uint64_t ROL64(const TypeID type_id, size_t nbits) {
...
-    }
-    return static_cast<TypeID>(TYPEID_INCOMPLETE_STRUCT_TYPE + hash % TYPEID_TYPE_RANGE);
-  } else {
-    for (const auto& f : s.fields) {
-      assert(f.type_id != TypeID::INVALID_TYPE);
-      hash ^= ROL64(f.type_id, i + 17u) ^ current::ROL64(current::CRC32(f.name), i + 29u);
-      ++i;
-    }
-    return static_cast<TypeID>(TYPEID_STRUCT_TYPE + hash % TYPEID_TYPE_RANGE);
+  hash ^= (static_cast<uint64_t>(s.super_id) ^ static_cast<uint64_t>(TypeID::CurrentStruct));
```
Comment (mzhurovich) [inline TypeSystem/Reflection/types.h]: No bit shift?

Comment (dkorolev) [inline TypeSystem/Reflection/types.h]: Shit, no bit shift. :+1: 1eb0200cdbaf92818e1723cd99f808eef40465d8

File: TypeSystem/Reflection/reflection.h (hunk 1 above)
Comment (dkorolev) [inline TypeSystem/Reflection/reflection.h]: Yes. :( 1eb0200cdbaf92818e1723cd99f808eef40465d8

**Comment by mzhurovich** [comment]
@dkorolev 
```
[ RUN      ] Reflection.StructAndVariant
./Reflection/test.cc:110: Failure
Value of: static_cast<uint64_t>(self_a.type_id)
  Actual: 9206142970491708268
Expected: 9205361767741727559ull
Which is: 9205361767741727559
./Reflection/test.cc:111: Failure
Value of: static_cast<uint64_t>(self_a.fields[0].type_id)
  Actual: 9319767815318439043
Expected: 9314504030810624379ull
Which is: 9314504030810624379
./Reflection/test.cc:114: Failure
Value of: static_cast<uint64_t>(self_b.type_id)
  Actual: 9206142973158716118
Expected: 9209719789868970707ull
Which is: 9209719789868970707
./Reflection/test.cc:115: Failure
Value of: static_cast<uint64_t>(self_b.fields[0].type_id)
  Actual: 9319767815318444419
Expected: 9319310138454761363ull
Which is: 9319310138454761363
./Reflection/test.cc:118: Failure
Value of: static_cast<uint64_t>(self_c.type_id)
  Actual: 9208204267910656623
Expected: 9200831696630769615ull
Which is: 9200831696630769615
./Reflection/test.cc:119: Failure
Value of: static_cast<uint64_t>(self_c.fields[0].type_id)
  Actual: 9319767815318444419
Expected: 9319310138454761363ull
Which is: 9319310138454761363
./Reflection/test.cc:120: Failure
Value of: static_cast<uint64_t>(self_c.fields[1].type_id)
  Actual: 9341629473358251575
Expected: 9344235412793316919ull
Which is: 9344235412793316919
[  FAILED  ] Reflection.StructAndVariant (18 ms)
```

**Comment by dkorolev** [comment]
@mzhurovich It's just there was no good code. Once good code is in, tests will pass. 
Colgar en allí, and have a nice day!

**Comment by dkorolev** [comment]
Oh, look, the `make check` fix turned out to be not hard at all too!

**Comment by sompylasar** [comment]
> Colgar en allí, and have a nice day!

Colgar allí ;)

**Pull request merged**